### PR TITLE
[improve][fn] Support e2e cryption in python instance 

### LINKED ISSUE #18626: Support e2e in python function
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Motivation

python instance doesn't support e2e cryption

### Solution

_No response_

### Alternatives

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by jiangpengcheng**
Fixes #18626

### Motivation

Python instance doesn't support e2e encryption yet

### Modifications

Add e2e encryption support

### Verifying this change

- [x] Make sure that the change passes the CI checks.

- [x] This change is already covered by existing tests, such as *(please describe tests)*.
  - *org.apache.pulsar.tests.integration.functions.python.PulsarFunctionsPythonTest*

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: https://github.com/jiangpengcheng/pulsar/pull/4

### CODE REVIEW

**Comment by nodece** [comment]
I don't think you should add the generic runtime support to this PR. Could you make a new PR to do that?

**Review by freeznet — APPROVED** [review]

File: pulsar-functions/instance/src/main/python/python_instance.py [hunk 1]
```diff
@@ -460,3 +472,16 @@ def close(self):
...
+    crypto_key_reader = None
+    if crypto_spec is not None:
+      try:
+        crypto_config = json.loads(crypto_spec.cryptoKeyReaderConfig)
+        if crypto_spec.cryptoKeyReaderClassName == "" or crypto_spec.cryptoKeyReaderClassName is None:
+          crypto_key_reader = pulsar.CryptoKeyReader(**crypto_config)
+        else:
+          crypto_key_reader = util.import_class(os.path.dirname(self.user_code), crypto_spec.cryptoKeyReaderClassName)(**crypto_config)
+      except:
+        pass
```
Comment (freeznet) [inline pulsar-functions/instance/src/main/python/python_instance.py:486]: we may need to print the exception to user.

Comment (nlu90) [reply pulsar-functions/instance/src/main/python/python_instance.py:486]: +1 for a clear error message

Comment (jiangpengcheng) [reply pulsar-functions/instance/src/main/python/python_instance.py:488]: I will update

**Review by nlu90 — APPROVED** [review]
@nodece @BewareMyPower Could you help also take a look?

**Review by BewareMyPower — APPROVED** [review]

**Comment by Anonymitaet** [comment]
Hi @jiangpengcheng I see this PR is labeled with `doc-not-needed`. Is this PR an internal improvement and does not affect end users? If no, do we need to add some usage instructions to https://pulsar.apache.org/docs/next/functions-develop-security/? TIA!

**Comment by jiangpengcheng** [comment]
@Anonymitaet 

yes, this is an internal improvement

- yes, we need to tick 2 boxes of `e-2-e` for python

- the `cryptoConfig` is passed via `--input-specs`, so there is no need to add a new flag to the reference list

- I see that the `function develop security` means to use secret data in the user's function, while this PR is to enable e2e encryption support on the input pulsar topics(producer sends encrypted data to pulsar, consumer received encrypted data from pulsar and decrypted it) for python runtime, there are different, so doesn't need to add examples to it

**Comment by Anonymitaet** [comment]
@jiangpengcheng OK, thank you! 

Done

>yes, we need to tick 2 boxes of e-2-e for python

**Pull request merged**